optlang/glpk_interface.py

```diff
@@ -585,45 +586,37 @@ def objective(self, value):
 
     @property
     def primal_values(self):
-        primal_values = collections.OrderedDict()
-        is_mip = self._glpk_is_mip()
-        for index, variable in enumerate(self.variables):
-            if is_mip:
-                value = glp_mip_col_val(self.problem, index + 1)
-            else:
-                value = glp_get_col_prim(self.problem, index + 1)
-            primal_values[variable.name] = variable._round_primal_to_bounds(value)
-        return primal_values
+        # round primals
+        primal_values = [variable._round_primal_to_bounds(primal)
+                         for variable, primal in zip(self.variables, self._get_primal_values())]
+        return collections.OrderedDict(
+            zip(self._get_variables_names(), primal_values)
+        )
 
-    @property
-    def reduced_costs(self):
+    def _get_primal_values(self):
+        if self._glpk_is_mip():
+            # no vector function (element wise)
+            return [glp_mip_col_val(self.problem, index + 1)
+                    for index in range(len(self.variables))]
+        else:
+            return get_col_primals(self.problem)
```

Comment:
get_col_primals will return the mip values via glp_mip_col_val automatically if the problem is a MIP. That is done in C so no need for the if-statement.